File: DSA_python/DSA01003 HOÁN VỊ KẾ TIẾP.py
def sort_list(A, start, end):
    B = []
    for i in range(start, end):
        B.append(A[i])

    B.sort()
    j = 0
    for i in range(start, end):
        A[i] = B[j]
        j += 1
    return A

def update(A, n):
    i = n-2
    while A[i] > A[i+1]:
        i -= 1

    if i != -1:
        k = n-1
        while A[k] < A[i]:
            k -= 1

        A[i], A[k] = A[k], A[i]
        A = sort_list(A, i+1, n)
    else:
        A.reverse()
    return A

def submit():
    n = int(input())
    A = list(map(int, input().split()))

    A = update(A, n)

    for i in A:
        print(i, end = " ")
    print()

test_num = int(input())
for i in range(test_num):
    submit()

File: DSA/DSA01003 HOÁN VỊ KẾ TIẾP.cpp
#include <bits/stdc++.h>
using namespace std;

void update(vector<int> &A, int n)
{
    int i = n-2;
    while (A[i] > A[i+1])
    {
        i--;
    }
    
    if (i != -1)
    {
        int k = n-1;
        while (A[k] < A[i])
        {
            k--;
        }
        
        swap(A[i], A[k]);
        sort(A.begin() + i+1, A.end());
    }
    else
    {
        reverse(A.begin(), A.end());
    }
    
}

void submit()
{
    int n;
    cin >> n;
    vector<int> A(n);
    for (auto &&i : A)
    {
        cin >> i;
    }
    
    update(A, n);
    for (auto &&i : A)
    {
        cout << i << " ";
    }
    cout << endl;
}   

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();

    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}